File: src/org/jabby/blazenav/BlazeNavToLineAction.java
package org.jabby.blazenav;

import com.google.common.collect.Lists;
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.CustomShortcutSet;
import com.intellij.openapi.editor.Editor;
import com.intellij.psi.PsiFile;
import com.intellij.util.Range;

import javax.swing.*;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class BlazeNavToLineAction extends NavActionBase {
    @Override void actionPerformed(PsiFile psiFile, Editor editor) {
        Runnable removeHighlightMaskFn = disableHighlighting(editor, psiFile.getTextLength());
        Runnable restoreFoldsFn = unfold(editor);

        Range<Integer> range = getVisibleLinesRange(editor);
        Map<Integer, Integer> linesOffsets = getLinesOffsets(psiFile.getText(), range.getFrom(), range.getTo());
        Map<Character, Integer> mappings = mapKeysToOffsets(linesOffsets);

        Runnable unhighlightCharsFn = highlightChars(editor, mappings);

        List<AnAction> nextActions = new ArrayList<AnAction>(linesOffsets.size());
        JComponent editorComponent = editor.getContentComponent();
        Runnable unregisterShortcutsFn = createUnregisterShortcutsFn(editorComponent, nextActions);

        for (Character c : mappings.keySet()) {
            MoveCaretToOffsetAction next =
                    new MoveCaretToOffsetAction(mappings.get(c),
                            removeHighlightMaskFn, restoreFoldsFn, unregisterShortcutsFn, unhighlightCharsFn);
            next.registerCustomShortcutSet(new CustomShortcutSet(createKeyStroke(c)), editorComponent);
            nextActions.add(next);
        }

        setupCancellation(
                editorComponent,
                unregisterShortcutsFn, removeHighlightMaskFn, restoreFoldsFn,
                unhighlightCharsFn, createClearStatusFn(editor));
    }

    Map<Integer, Integer> getLinesOffsets(String text, int fromLine, int toLine) {
        int fromOffset = getFromOffset(text, fromLine);
        int toOffset = getToOffset(text, toLine);

        Pattern pattern = Pattern.compile("\\n");
        Matcher matcher = pattern.matcher(text.substring(fromOffset, toOffset));

        Map<Integer, Integer> map = new HashMap<Integer, Integer>(toLine - fromLine);
        map.put(fromLine, fromOffset);
        int i = 1;
        while (matcher.find()) {
            map.put(i++ + fromLine, matcher.start() + fromOffset + 1);
        }
        return map;
    }

    Map<Character, Integer> mapKeysToOffsets(Map<Integer, Integer> linesOffsets) {
        List<Integer> affectedLines =
                Lists.newArrayList(linesOffsets.keySet())
                        .subList(0, Math.min(linesOffsets.size(), SUBSTITUTIONS.length));
        Collections.sort(affectedLines);
        Map<Character, Integer> mappings = new HashMap<Character, Integer>(affectedLines.size());
        int i = 0;
        for (int line : affectedLines) { mappings.put(SUBSTITUTIONS[i++], linesOffsets.get(line)); }
        return mappings;
    }
}

File: src/org/jabby/blazenav/BlazeNavToLineAction.kt
package org.jabby.blazenav

import com.intellij.openapi.editor.Editor
import org.jabby.blazenav.NavActionBase.Bounds
import java.util.regex.Pattern
import java.util.LinkedHashMap
import java.util.HashMap
import java.util.ArrayList
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.CustomShortcutSet

public class BlazeNavToLineAction: NavActionBase() {
    protected override fun actionPerformed(editor: Editor) {
        val text = editor.getDocument().getText()!!

        val unFadeFn = fade(editor, text.length())
        val unUnfoldFn = unfold(editor)

        val bounds = getVisibleBounds(editor)
        val lineBeginningsOffsets = getLineBeginningsOffsets(text, bounds)
        val keyToOffsetsMappings = getKeyToOffsetsMappings(lineBeginningsOffsets)

        val unHighlightCharsFn = highlightChars(editor, keyToOffsetsMappings)

        val nextActions: MutableList<AnAction> = ArrayList(lineBeginningsOffsets.size)
        val editorComponent = editor.getContentComponent()
        val unRegisterShortcutsFn = createUnRegisterShortcutsFn(editorComponent, nextActions)

        for ((char, offset) in keyToOffsetsMappings) {
            val next =
                    MoveCaretToOffsetAction(keyToOffsetsMappings.get(char)!!,
                            unFadeFn, unUnfoldFn, unRegisterShortcutsFn, unHighlightCharsFn)
            next.registerCustomShortcutSet(CustomShortcutSet(createKeyStroke(char)), editorComponent)
            nextActions.add(next)
        }

        setupCancellation(
                editorComponent,
                unFadeFn, unUnfoldFn, unRegisterShortcutsFn, unHighlightCharsFn, createClearStatusFn(editor)
        )
    }

    private fun getLineBeginningsOffsets(text: String, bounds: Bounds): Map<Int, Int> {
        val fromOffset = getFromOffset(text, bounds.fromLine)
        val toOffset = getToOffset(text, bounds.toLine)

        val visibleText = text.subSequence(fromOffset, toOffset)
        val matcher = Pattern.compile("\\n").matcher(visibleText)

        val map: MutableMap<Int, Int> = LinkedHashMap(bounds.toLine - bounds.fromLine)
        map.put(bounds.fromLine, fromOffset)

        var i = 1
        while (matcher.find()) map.put(i++ + bounds.fromLine, matcher.start() + fromOffset + 1)
        return map
    }

    private fun getKeyToOffsetsMappings(lineBeginningsOffsets: Map<Int, Int>): Map<Char, Int> {
        val lines =
                lineBeginningsOffsets.keySet().take(
                        Math.min(lineBeginningsOffsets.size, SUBSTITUTIONS.size))
        val mappings: MutableMap<Char, Int> = HashMap(lines.size)
        var i = 0
        lines.forEach { mappings.put(SUBSTITUTIONS[i++], lineBeginningsOffsets.get(it)!!) }
        return mappings
    }
}
